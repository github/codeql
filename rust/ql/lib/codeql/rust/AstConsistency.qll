/**
 * Provides classes for recognizing control flow graph inconsistencies.
 */

private import rust
private import codeql.rust.elements.internal.generated.ParentChild

private predicate multipleToStrings(Element e) { strictcount(e.toString()) > 1 }

/**
 * Holds if `e` has more than one `toString()` result.
 */
query predicate multipleToStrings(Element e, string s) {
  multipleToStrings(e) and
  s = strictconcat(e.toString(), ", ")
}

/**
 * Holds if `e` has more than one `Location`.
 */
query predicate multipleLocations(Locatable e) { strictcount(e.getLocation()) > 1 }

private predicate multiplePrimaryQlClasses(Element e) { strictcount(e.getAPrimaryQlClass()) > 1 }

/**
 * Holds if `e` has more than one `getPrimaryQlClasses()` result.
 */
query predicate multiplePrimaryQlClasses(Element e, string s) {
  multiplePrimaryQlClasses(e) and
  s = strictconcat(e.getPrimaryQlClasses(), ", ")
}

private Element getParent(Element child) { child = getChildAndAccessor(result, _, _) }

private predicate multipleParents(Element child) { strictcount(getParent(child)) > 1 }

/**
 * Holds if `child` has more than one AST parent.
 */
query predicate multipleParents(Element child, Element parent) {
  multipleParents(child) and
  parent = getParent(child)
}

/**
 * Gets counts of abstract syntax tree inconsistencies of each type.
 */
int getAstInconsistencyCounts(string type) {
  // total results from all the AST consistency query predicates.
  type = "Multiple toStrings" and
  result = count(Element e | multipleToStrings(e) | e)
  or
  type = "Multiple locations" and
  result = count(Element e | multipleLocations(e) | e)
  or
  type = "Multiple primary QL classes" and
  result = count(Element e | multiplePrimaryQlClasses(e) | e)
  or
  type = "Multiple parents" and
  result = count(Element e | multipleParents(e) | e)
}
