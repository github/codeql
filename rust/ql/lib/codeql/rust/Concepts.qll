/**
 * Provides abstract classes representing generic concepts such as file system
 * access or system command execution, for which individual framework libraries
 * provide concrete subclasses.
 */

private import codeql.rust.dataflow.DataFlow
private import codeql.threatmodels.ThreatModels
private import codeql.rust.Frameworks

/**
 * A data flow source for a specific threat-model.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `ThreatModelSource::Range` instead.
 */
final class ThreatModelSource = ThreatModelSource::Range;

/**
 * Provides a class for modeling new sources for specific threat-models.
 */
module ThreatModelSource {
  /**
   * A data flow source, for a specific threat-model.
   */
  abstract class Range extends DataFlow::Node {
    /**
     * Gets a string that represents the source kind with respect to threat modeling.
     *
     * See
     * - https://github.com/github/codeql/blob/main/docs/codeql/reusables/threat-model-description.rst
     * - https://github.com/github/codeql/blob/main/shared/threat-models/ext/threat-model-grouping.model.yml
     */
    abstract string getThreatModel();

    /**
     * Gets a string that describes the type of this threat-model source.
     */
    abstract string getSourceType();
  }
}

/**
 * A data flow source that is enabled in the current threat model configuration.
 */
class ActiveThreatModelSource extends ThreatModelSource {
  ActiveThreatModelSource() { currentThreatModel(this.getThreatModel()) }
}

/**
 * A data flow source corresponding to the program's command line arguments or path.
 */
final class CommandLineArgsSource = CommandLineArgsSource::Range;

/**
 * Provides a class for modeling new sources for the program's command line arguments or path.
 */
module CommandLineArgsSource {
  /**
   * A data flow source corresponding to the program's command line arguments or path.
   */
  abstract class Range extends ThreatModelSource::Range {
    override string getThreatModel() { result = "commandargs" }

    override string getSourceType() { result = "CommandLineArgs" }
  }
}

/**
 * A data flow source corresponding to the program's environment.
 */
final class EnvironmentSource = EnvironmentSource::Range;

/**
 * Provides a class for modeling new sources for the program's environment.
 */
module EnvironmentSource {
  /**
   * A data flow source corresponding to the program's environment.
   */
  abstract class Range extends ThreatModelSource::Range {
    override string getThreatModel() { result = "environment" }

    override string getSourceType() { result = "EnvironmentSource" }
  }
}

/**
 * A data flow source for remote (network) data.
 */
final class RemoteSource = RemoteSource::Range;

/**
 * Provides a class for modeling new sources of remote (network) data.
 */
module RemoteSource {
  /**
   * A data flow source for remote (network) data.
   */
  abstract class Range extends ThreatModelSource::Range {
    override string getThreatModel() { result = "remote" }

    override string getSourceType() { result = "RemoteSource" }
  }
}

/**
 * A data flow node that constructs a SQL statement (for later execution).
 *
 * Often, it is worthy of an alert if a SQL statement is constructed such that
 * executing it would be a security risk.
 *
 * If it is important that the SQL statement is executed, use `SqlExecution`.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `SqlConstruction::Range` instead.
 */
final class SqlConstruction = SqlConstruction::Range;

/**
 * Provides a class for modeling new SQL execution APIs.
 */
module SqlConstruction {
  /**
   * A data flow node that constructs a SQL statement.
   */
  abstract class Range extends DataFlow::Node {
    /**
     * Gets the argument that specifies the SQL statements to be constructed.
     */
    abstract DataFlow::Node getSql();
  }
}

/**
 * A data flow node that constructs and executes SQL statements.
 *
 * If the context of interest is such that merely constructing a SQL statement
 * would be valuable to report, consider also using `SqlConstruction`.
 *
 * Extend this class to refine existing API models. If you want to model new APIs,
 * extend `SqlExecution::Range` instead.
 */
final class SqlExecution = SqlExecution::Range;

/**
 * Provides a class for modeling new SQL execution APIs.
 */
module SqlExecution {
  /**
   * A data flow node that executes SQL statements.
   */
  abstract class Range extends DataFlow::Node {
    /**
     * Gets the argument that specifies the SQL statements to be executed.
     */
    abstract DataFlow::Node getSql();
  }
}

/**
 * A data flow node that performs SQL sanitization.
 */
final class SqlSanitization = SqlSanitization::Range;

/**
 * Provides a class for modeling new SQL sanitization APIs.
 */
module SqlSanitization {
  /**
   * A data flow node that performs SQL sanitization.
   */
  abstract class Range extends DataFlow::Node { }
}

/**
 * Provides models for cryptographic things.
 */
module Cryptography {
  private import codeql.rust.internal.ConceptsShared::Cryptography as SC

  final class CryptographicOperation = SC::CryptographicOperation;

  class EncryptionAlgorithm = SC::EncryptionAlgorithm;

  class HashingAlgorithm = SC::HashingAlgorithm;

  class PasswordHashingAlgorithm = SC::PasswordHashingAlgorithm;

  module CryptographicOperation = SC::CryptographicOperation;

  class BlockMode = SC::BlockMode;

  class CryptographicAlgorithm = SC::CryptographicAlgorithm;
}
