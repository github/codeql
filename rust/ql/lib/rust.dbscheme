// generated by codegen, do not edit

// from ../shared/tree-sitter-extractor/src/generator/prefix.dbscheme
/*- Files and folders -*/

/**
 * The location of an element.
 * The location spans column `startcolumn` of line `startline` to
 * column `endcolumn` of line `endline` in file `file`.
 * For more information, see
 * [Locations](https://codeql.github.com/docs/writing-codeql-queries/providing-locations-in-codeql-queries/).
 */
locations_default(
  unique int id: @location_default,
  int file: @file ref,
  int beginLine: int ref,
  int beginColumn: int ref,
  int endLine: int ref,
  int endColumn: int ref
);

files(
  unique int id: @file,
  string name: string ref
);

folders(
  unique int id: @folder,
  string name: string ref
);

@container = @file | @folder

containerparent(
  int parent: @container ref,
  unique int child: @container ref
);

/*- Empty location -*/

empty_location(
  int location: @location_default ref
);

/*- Source location prefix -*/

/**
 * The source location of the snapshot.
 */
sourceLocationPrefix(string prefix : string ref);

/*- Diagnostic messages -*/

diagnostics(
  unique int id: @diagnostic,
  int severity: int ref,
  string error_tag: string ref,
  string error_message: string ref,
  string full_error_message: string ref,
  int location: @location_default ref
);

/*- Diagnostic messages: severity -*/

case @diagnostic.severity of
  10 = @diagnostic_debug
| 20 = @diagnostic_info
| 30 = @diagnostic_warning
| 40 = @diagnostic_error
;

/*- YAML -*/

#keyset[parent, idx]
yaml (unique int id: @yaml_node,
      int kind: int ref,
      int parent: @yaml_node_parent ref,
      int idx: int ref,
      string tag: string ref,
      string tostring: string ref);

case @yaml_node.kind of
  0 = @yaml_scalar_node
| 1 = @yaml_mapping_node
| 2 = @yaml_sequence_node
| 3 = @yaml_alias_node
;

@yaml_collection_node = @yaml_mapping_node | @yaml_sequence_node;

@yaml_node_parent = @yaml_collection_node | @file;

yaml_anchors (unique int node: @yaml_node ref,
              string anchor: string ref);

yaml_aliases (unique int alias: @yaml_alias_node ref,
              string target: string ref);

yaml_scalars (unique int scalar: @yaml_scalar_node ref,
              int style: int ref,
              string value: string ref);

yaml_errors (unique int id: @yaml_error,
             string message: string ref);

yaml_locations(unique int locatable: @yaml_locatable ref,
             int location: @location_default ref);

@yaml_locatable = @yaml_node | @yaml_error;


// from prefix.dbscheme
#keyset[id]
locatable_locations(
  int id: @locatable ref,
  int location: @location_default ref
);


// from schema.py

@element =
  @locatable
| @unextracted
;

@locatable =
  @ast_node
;

@unextracted =
  @missing
| @unimplemented
;

@ast_node =
  @declaration
| @expr
| @generic_arg_list
| @label
| @match_arm
| @pat
| @path
| @record_expr_field
| @record_pat_field
| @stmt
| @type_ref
;

@missing =
  @missing_expr
| @missing_pat
;

@unimplemented =
  @generic_arg_list
| @path
| @type_ref
| @unimplemented_declaration
;

@declaration =
  @function
| @module
| @unimplemented_declaration
;

@expr =
  @array_expr
| @asm_expr
| @await_expr
| @become_expr
| @binary_expr
| @block_expr_base
| @box_expr
| @break_expr
| @call_expr
| @cast_expr
| @closure_expr
| @const_expr
| @continue_expr
| @field_expr
| @if_expr
| @index_expr
| @let_expr
| @literal_expr
| @loop_expr
| @match_expr
| @method_call_expr
| @missing_expr
| @offset_of_expr
| @path_expr
| @prefix_expr
| @range_expr
| @record_expr
| @ref_expr
| @return_expr
| @tuple_expr
| @underscore_expr
| @yeet_expr
| @yield_expr
;

generic_arg_lists(
  unique int id: @generic_arg_list
);

labels(
  unique int id: @label,
  string name: string ref
);

match_arms(
  unique int id: @match_arm,
  int pat: @pat ref,
  int expr: @expr ref
);

#keyset[id]
match_arm_guards(
  int id: @match_arm ref,
  int guard: @expr ref
);

@pat =
  @box_pat
| @const_block_pat
| @ident_pat
| @literal_pat
| @missing_pat
| @or_pat
| @path_pat
| @range_pat
| @record_pat
| @ref_pat
| @slice_pat
| @tuple_pat
| @tuple_struct_pat
| @wildcard_pat
;

paths(
  unique int id: @path
);

record_expr_fields(
  unique int id: @record_expr_field,
  string name: string ref,
  int expr: @expr ref
);

record_pat_fields(
  unique int id: @record_pat_field,
  string name: string ref,
  int pat: @pat ref
);

@stmt =
  @expr_stmt
| @item_stmt
| @let_stmt
;

type_refs(
  unique int id: @type_ref
);

@array_expr =
  @element_list_expr
| @repeat_expr
;

asm_exprs(
  unique int id: @asm_expr,
  int expr: @expr ref
);

await_exprs(
  unique int id: @await_expr,
  int expr: @expr ref
);

become_exprs(
  unique int id: @become_expr,
  int expr: @expr ref
);

binary_exprs(
  unique int id: @binary_expr,
  int lhs: @expr ref,
  int rhs: @expr ref
);

#keyset[id]
binary_expr_ops(
  int id: @binary_expr ref,
  string op: string ref
);

@block_expr_base =
  @async_block_expr
| @block_expr
| @unsafe_block_expr
;

#keyset[id, index]
block_expr_base_statements(
  int id: @block_expr_base ref,
  int index: int ref,
  int statement: @stmt ref
);

#keyset[id]
block_expr_base_tails(
  int id: @block_expr_base ref,
  int tail: @expr ref
);

box_exprs(
  unique int id: @box_expr,
  int expr: @expr ref
);

box_pats(
  unique int id: @box_pat,
  int inner: @pat ref
);

break_exprs(
  unique int id: @break_expr
);

#keyset[id]
break_expr_exprs(
  int id: @break_expr ref,
  int expr: @expr ref
);

#keyset[id]
break_expr_labels(
  int id: @break_expr ref,
  int label: @label ref
);

call_exprs(
  unique int id: @call_expr,
  int callee: @expr ref
);

#keyset[id, index]
call_expr_args(
  int id: @call_expr ref,
  int index: int ref,
  int arg: @expr ref
);

#keyset[id]
call_expr_is_assignee_expr(
  int id: @call_expr ref
);

cast_exprs(
  unique int id: @cast_expr,
  int expr: @expr ref,
  int type_: @type_ref ref
);

closure_exprs(
  unique int id: @closure_expr,
  int body: @expr ref,
  string closure_kind: string ref
);

#keyset[id, index]
closure_expr_args(
  int id: @closure_expr ref,
  int index: int ref,
  int arg: @pat ref
);

#keyset[id, index]
closure_expr_arg_types(
  int id: @closure_expr ref,
  int index: int ref,
  int arg_type: @type_ref ref
);

#keyset[id]
closure_expr_ret_types(
  int id: @closure_expr ref,
  int ret_type: @type_ref ref
);

#keyset[id]
closure_expr_is_move(
  int id: @closure_expr ref
);

const_block_pats(
  unique int id: @const_block_pat,
  int expr: @expr ref
);

const_exprs(
  unique int id: @const_expr,
  int expr: @expr ref
);

continue_exprs(
  unique int id: @continue_expr
);

#keyset[id]
continue_expr_labels(
  int id: @continue_expr ref,
  int label: @label ref
);

expr_stmts(
  unique int id: @expr_stmt,
  int expr: @expr ref
);

#keyset[id]
expr_stmt_has_semicolon(
  int id: @expr_stmt ref
);

field_exprs(
  unique int id: @field_expr,
  int expr: @expr ref,
  string name: string ref
);

functions(
  unique int id: @function,
  string name: string ref,
  int body: @expr ref
);

ident_pats(
  unique int id: @ident_pat,
  string binding_id: string ref
);

#keyset[id]
ident_pat_subpats(
  int id: @ident_pat ref,
  int subpat: @pat ref
);

if_exprs(
  unique int id: @if_expr,
  int condition: @expr ref,
  int then: @expr ref
);

#keyset[id]
if_expr_elses(
  int id: @if_expr ref,
  int else: @expr ref
);

index_exprs(
  unique int id: @index_expr,
  int base: @expr ref,
  int index: @expr ref
);

#keyset[id]
index_expr_is_assignee_expr(
  int id: @index_expr ref
);

item_stmts(
  unique int id: @item_stmt
);

let_exprs(
  unique int id: @let_expr,
  int pat: @pat ref,
  int expr: @expr ref
);

let_stmts(
  unique int id: @let_stmt,
  int pat: @pat ref
);

#keyset[id]
let_stmt_types(
  int id: @let_stmt ref,
  int type_: @type_ref ref
);

#keyset[id]
let_stmt_initializers(
  int id: @let_stmt ref,
  int initializer: @expr ref
);

#keyset[id]
let_stmt_elses(
  int id: @let_stmt ref,
  int else: @expr ref
);

literal_exprs(
  unique int id: @literal_expr
);

literal_pats(
  unique int id: @literal_pat,
  int expr: @expr ref
);

loop_exprs(
  unique int id: @loop_expr,
  int body: @expr ref
);

#keyset[id]
loop_expr_labels(
  int id: @loop_expr ref,
  int label: @label ref
);

match_exprs(
  unique int id: @match_expr,
  int expr: @expr ref
);

#keyset[id, index]
match_expr_branches(
  int id: @match_expr ref,
  int index: int ref,
  int branch: @match_arm ref
);

method_call_exprs(
  unique int id: @method_call_expr,
  int receiver: @expr ref,
  string method_name: string ref
);

#keyset[id, index]
method_call_expr_args(
  int id: @method_call_expr ref,
  int index: int ref,
  int arg: @expr ref
);

#keyset[id]
method_call_expr_generic_args(
  int id: @method_call_expr ref,
  int generic_args: @generic_arg_list ref
);

missing_exprs(
  unique int id: @missing_expr
);

missing_pats(
  unique int id: @missing_pat
);

modules(
  unique int id: @module
);

#keyset[id, index]
module_declarations(
  int id: @module ref,
  int index: int ref,
  int declaration: @declaration ref
);

offset_of_exprs(
  unique int id: @offset_of_expr,
  int container: @type_ref ref
);

#keyset[id, index]
offset_of_expr_fields(
  int id: @offset_of_expr ref,
  int index: int ref,
  string field: string ref
);

or_pats(
  unique int id: @or_pat
);

#keyset[id, index]
or_pat_args(
  int id: @or_pat ref,
  int index: int ref,
  int arg: @pat ref
);

path_exprs(
  unique int id: @path_expr,
  int path: @path ref
);

path_pats(
  unique int id: @path_pat,
  int path: @path ref
);

prefix_exprs(
  unique int id: @prefix_expr,
  int expr: @expr ref,
  string op: string ref
);

range_exprs(
  unique int id: @range_expr
);

#keyset[id]
range_expr_lhs(
  int id: @range_expr ref,
  int lhs: @expr ref
);

#keyset[id]
range_expr_rhs(
  int id: @range_expr ref,
  int rhs: @expr ref
);

#keyset[id]
range_expr_is_inclusive(
  int id: @range_expr ref
);

range_pats(
  unique int id: @range_pat
);

#keyset[id]
range_pat_starts(
  int id: @range_pat ref,
  int start: @pat ref
);

#keyset[id]
range_pat_ends(
  int id: @range_pat ref,
  int end: @pat ref
);

record_exprs(
  unique int id: @record_expr
);

#keyset[id]
record_expr_paths(
  int id: @record_expr ref,
  int path: @path ref
);

#keyset[id, index]
record_expr_flds(
  int id: @record_expr ref,
  int index: int ref,
  int fld: @record_expr_field ref
);

#keyset[id]
record_expr_spreads(
  int id: @record_expr ref,
  int spread: @expr ref
);

#keyset[id]
record_expr_has_ellipsis(
  int id: @record_expr ref
);

#keyset[id]
record_expr_is_assignee_expr(
  int id: @record_expr ref
);

record_pats(
  unique int id: @record_pat
);

#keyset[id]
record_pat_paths(
  int id: @record_pat ref,
  int path: @path ref
);

#keyset[id, index]
record_pat_flds(
  int id: @record_pat ref,
  int index: int ref,
  int fld: @record_pat_field ref
);

#keyset[id]
record_pat_has_ellipsis(
  int id: @record_pat ref
);

ref_exprs(
  unique int id: @ref_expr,
  int expr: @expr ref
);

#keyset[id]
ref_expr_is_raw(
  int id: @ref_expr ref
);

#keyset[id]
ref_expr_is_mut(
  int id: @ref_expr ref
);

ref_pats(
  unique int id: @ref_pat,
  int pat: @pat ref
);

#keyset[id]
ref_pat_is_mut(
  int id: @ref_pat ref
);

return_exprs(
  unique int id: @return_expr
);

#keyset[id]
return_expr_exprs(
  int id: @return_expr ref,
  int expr: @expr ref
);

slice_pats(
  unique int id: @slice_pat
);

#keyset[id, index]
slice_pat_prefixes(
  int id: @slice_pat ref,
  int index: int ref,
  int prefix: @pat ref
);

#keyset[id]
slice_pat_slice(
  int id: @slice_pat ref,
  int slice: @pat ref
);

#keyset[id, index]
slice_pat_suffixes(
  int id: @slice_pat ref,
  int index: int ref,
  int suffix: @pat ref
);

tuple_exprs(
  unique int id: @tuple_expr
);

#keyset[id, index]
tuple_expr_exprs(
  int id: @tuple_expr ref,
  int index: int ref,
  int expr: @expr ref
);

#keyset[id]
tuple_expr_is_assignee_expr(
  int id: @tuple_expr ref
);

tuple_pats(
  unique int id: @tuple_pat
);

#keyset[id, index]
tuple_pat_args(
  int id: @tuple_pat ref,
  int index: int ref,
  int arg: @pat ref
);

#keyset[id]
tuple_pat_ellipsis_indices(
  int id: @tuple_pat ref,
  int ellipsis_index: int ref
);

tuple_struct_pats(
  unique int id: @tuple_struct_pat
);

#keyset[id]
tuple_struct_pat_paths(
  int id: @tuple_struct_pat ref,
  int path: @path ref
);

#keyset[id, index]
tuple_struct_pat_args(
  int id: @tuple_struct_pat ref,
  int index: int ref,
  int arg: @pat ref
);

#keyset[id]
tuple_struct_pat_ellipsis_indices(
  int id: @tuple_struct_pat ref,
  int ellipsis_index: int ref
);

underscore_exprs(
  unique int id: @underscore_expr
);

unimplemented_declarations(
  unique int id: @unimplemented_declaration
);

wildcard_pats(
  unique int id: @wildcard_pat
);

yeet_exprs(
  unique int id: @yeet_expr
);

#keyset[id]
yeet_expr_exprs(
  int id: @yeet_expr ref,
  int expr: @expr ref
);

yield_exprs(
  unique int id: @yield_expr
);

#keyset[id]
yield_expr_exprs(
  int id: @yield_expr ref,
  int expr: @expr ref
);

async_block_exprs(
  unique int id: @async_block_expr
);

block_exprs(
  unique int id: @block_expr
);

#keyset[id]
block_expr_labels(
  int id: @block_expr ref,
  int label: @label ref
);

element_list_exprs(
  unique int id: @element_list_expr
);

#keyset[id, index]
element_list_expr_elements(
  int id: @element_list_expr ref,
  int index: int ref,
  int element: @expr ref
);

#keyset[id]
element_list_expr_is_assignee_expr(
  int id: @element_list_expr ref
);

repeat_exprs(
  unique int id: @repeat_expr,
  int initializer: @expr ref,
  int repeat: @expr ref
);

unsafe_block_exprs(
  unique int id: @unsafe_block_expr
);
