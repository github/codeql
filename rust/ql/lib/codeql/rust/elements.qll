// generated by codegen, do not edit
/**
 * This module exports all modules providing `Element` subclasses.
 */

import codeql.rust.elements.Abi
import codeql.rust.elements.ArgList
import codeql.rust.elements.ArrayExpr
import codeql.rust.elements.ArrayType
import codeql.rust.elements.AsmExpr
import codeql.rust.elements.AssocItem
import codeql.rust.elements.AssocItemList
import codeql.rust.elements.AssocTypeArg
import codeql.rust.elements.AstNode
import codeql.rust.elements.Attr
import codeql.rust.elements.AwaitExpr
import codeql.rust.elements.BecomeExpr
import codeql.rust.elements.BinaryExpr
import codeql.rust.elements.BlockExpr
import codeql.rust.elements.BoxPat
import codeql.rust.elements.BreakExpr
import codeql.rust.elements.CallExpr
import codeql.rust.elements.CastExpr
import codeql.rust.elements.ClosureBinder
import codeql.rust.elements.ClosureExpr
import codeql.rust.elements.Const
import codeql.rust.elements.ConstArg
import codeql.rust.elements.ConstBlockPat
import codeql.rust.elements.ConstParam
import codeql.rust.elements.ContinueExpr
import codeql.rust.elements.DynTraitType
import codeql.rust.elements.Element
import codeql.rust.elements.Enum
import codeql.rust.elements.Expr
import codeql.rust.elements.ExprStmt
import codeql.rust.elements.ExternBlock
import codeql.rust.elements.ExternCrate
import codeql.rust.elements.ExternItem
import codeql.rust.elements.ExternItemList
import codeql.rust.elements.FieldExpr
import codeql.rust.elements.FieldList
import codeql.rust.elements.FnPtrType
import codeql.rust.elements.ForExpr
import codeql.rust.elements.ForType
import codeql.rust.elements.FormatArgsArg
import codeql.rust.elements.FormatArgsExpr
import codeql.rust.elements.Function
import codeql.rust.elements.GenericArg
import codeql.rust.elements.GenericArgList
import codeql.rust.elements.GenericParam
import codeql.rust.elements.GenericParamList
import codeql.rust.elements.IdentPat
import codeql.rust.elements.IfExpr
import codeql.rust.elements.Impl
import codeql.rust.elements.ImplTraitType
import codeql.rust.elements.IndexExpr
import codeql.rust.elements.InferType
import codeql.rust.elements.Item
import codeql.rust.elements.ItemList
import codeql.rust.elements.Label
import codeql.rust.elements.LetElse
import codeql.rust.elements.LetExpr
import codeql.rust.elements.LetStmt
import codeql.rust.elements.Lifetime
import codeql.rust.elements.LifetimeArg
import codeql.rust.elements.LifetimeParam
import codeql.rust.elements.LiteralExpr
import codeql.rust.elements.LiteralPat
import codeql.rust.elements.Locatable
import codeql.rust.elements.LoopExpr
import codeql.rust.elements.MacroCall
import codeql.rust.elements.MacroDef
import codeql.rust.elements.MacroExpr
import codeql.rust.elements.MacroPat
import codeql.rust.elements.MacroRules
import codeql.rust.elements.MacroType
import codeql.rust.elements.MatchArm
import codeql.rust.elements.MatchArmList
import codeql.rust.elements.MatchExpr
import codeql.rust.elements.MatchGuard
import codeql.rust.elements.Meta
import codeql.rust.elements.MethodCallExpr
import codeql.rust.elements.Missing
import codeql.rust.elements.Module
import codeql.rust.elements.Name
import codeql.rust.elements.NameRef
import codeql.rust.elements.NeverType
import codeql.rust.elements.OffsetOfExpr
import codeql.rust.elements.OrPat
import codeql.rust.elements.Param
import codeql.rust.elements.ParamList
import codeql.rust.elements.ParenExpr
import codeql.rust.elements.ParenPat
import codeql.rust.elements.ParenType
import codeql.rust.elements.Pat
import codeql.rust.elements.Path
import codeql.rust.elements.PathExpr
import codeql.rust.elements.PathPat
import codeql.rust.elements.PathSegment
import codeql.rust.elements.PathType
import codeql.rust.elements.PrefixExpr
import codeql.rust.elements.PtrType
import codeql.rust.elements.RangeExpr
import codeql.rust.elements.RangePat
import codeql.rust.elements.RecordExpr
import codeql.rust.elements.RecordExprField
import codeql.rust.elements.RecordExprFieldList
import codeql.rust.elements.RecordField
import codeql.rust.elements.RecordFieldList
import codeql.rust.elements.RecordPat
import codeql.rust.elements.RecordPatField
import codeql.rust.elements.RecordPatFieldList
import codeql.rust.elements.RefExpr
import codeql.rust.elements.RefPat
import codeql.rust.elements.RefType
import codeql.rust.elements.Rename
import codeql.rust.elements.RestPat
import codeql.rust.elements.RetType
import codeql.rust.elements.ReturnExpr
import codeql.rust.elements.ReturnTypeSyntax
import codeql.rust.elements.SelfParam
import codeql.rust.elements.SlicePat
import codeql.rust.elements.SliceType
import codeql.rust.elements.SourceFile
import codeql.rust.elements.Static
import codeql.rust.elements.Stmt
import codeql.rust.elements.StmtList
import codeql.rust.elements.Struct
import codeql.rust.elements.TokenTree
import codeql.rust.elements.Trait
import codeql.rust.elements.TraitAlias
import codeql.rust.elements.TryExpr
import codeql.rust.elements.TupleExpr
import codeql.rust.elements.TupleField
import codeql.rust.elements.TupleFieldList
import codeql.rust.elements.TuplePat
import codeql.rust.elements.TupleStructPat
import codeql.rust.elements.TupleType
import codeql.rust.elements.TypeAlias
import codeql.rust.elements.TypeArg
import codeql.rust.elements.TypeBound
import codeql.rust.elements.TypeBoundList
import codeql.rust.elements.TypeParam
import codeql.rust.elements.TypeRef
import codeql.rust.elements.UnderscoreExpr
import codeql.rust.elements.Unextracted
import codeql.rust.elements.Unimplemented
import codeql.rust.elements.Union
import codeql.rust.elements.Use
import codeql.rust.elements.UseTree
import codeql.rust.elements.UseTreeList
import codeql.rust.elements.Variant
import codeql.rust.elements.VariantList
import codeql.rust.elements.Visibility
import codeql.rust.elements.WhereClause
import codeql.rust.elements.WherePred
import codeql.rust.elements.WhileExpr
import codeql.rust.elements.WildcardPat
import codeql.rust.elements.YeetExpr
import codeql.rust.elements.YieldExpr
