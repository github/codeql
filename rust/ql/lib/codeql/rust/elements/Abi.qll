// generated by codegen, do not edit
/**
 * This module provides the public class `Abi`.
 */

private import internal.AbiImpl
import codeql.rust.elements.AstNode

/**
 * A Abi. For example:
 * ```rust
 * todo!()
 * ```
 */
final class Abi = Impl::Abi;
