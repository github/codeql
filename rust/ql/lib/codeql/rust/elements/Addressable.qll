// generated by codegen, do not edit
/**
 * This module provides the public class `Addressable`.
 */

private import internal.AddressableImpl
import codeql.rust.elements.AstNode

/**
 * Something that can be addressed by a path.
 *
 * TODO: This does not yet include all possible cases.
 */
final class Addressable = Impl::Addressable;
