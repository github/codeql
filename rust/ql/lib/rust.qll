/** Top-level import for the Rust language pack */

import codeql.rust.elements
import codeql.Locations
import codeql.files.FileSystem
import codeql.rust.elements.LogicalOperation
