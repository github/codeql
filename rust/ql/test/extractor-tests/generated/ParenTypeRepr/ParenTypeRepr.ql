// generated by codegen, do not edit
import codeql.rust.elements
import TestUtils

query predicate instances(ParenTypeRepr x) { toBeTested(x) and not x.isUnknown() }

query predicate getTypeRepr(ParenTypeRepr x, TypeRepr getTypeRepr) {
  toBeTested(x) and not x.isUnknown() and getTypeRepr = x.getTypeRepr()
}
