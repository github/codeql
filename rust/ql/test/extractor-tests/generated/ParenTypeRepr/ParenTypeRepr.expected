instances
| gen_paren_type_repr.rs:7:12:7:16 | (i32) |
getTypeRepr
| gen_paren_type_repr.rs:7:12:7:16 | (i32) | gen_paren_type_repr.rs:7:13:7:15 | i32 |
